### PR TITLE
make haggadah with bracha using dsl

### LINKED ISSUE #53: Create a haggadah with a bracha using the dsl
When the user creates a Haggadah and views it from the dashboard, they will see a Haggadah which contains a bracha.

## Acceptance criteria
```gherkin
Given a user at their dashboard 
When the user clicks the create Haggadah button, submits the form, returns to the dashboard, and views the Haggadah
Then they will see a Haggadah that contains a bracha with a title and content
```
The appearance of the haggadah should mimic the photo below.
image

### LINKED ISSUE #54: Add test fixture which removes the data in the firestore emulator
After running the test, the data in the firestore emulator will be removed so the following test operates with an empty database.

### PR DESCRIPTION
**Pull request by ebarylko**
## Summary
Changed the Haggadah creation feature. The user now creates a Haggadah by choosing a title and a template (There is only one for now)

closes #53
closes #54

### src/haggadah/dsl.cljs
* Added file with functions for creating and parsing a Haggadah

### src/haggadah/events.cljs
* Changed `fetch-haggadot` handler so that it queries the haggadot ordered by timestamp (newest first)
* Added `query!` effect which runs a query against firestore with an `orderBy` modifier
* Changed `add-haggadah` to return `add-haggadah!` effect with the default Haggadah
* Changed `set-haggadah` to store the hiccup representation of the argument in the re-frame database 

### src/haggadah/views.cljs
* Changed `haggadah-creation-panel` so the user can only select the title of the Haggadah and which template they want to use
* Added dropdown menu which shows the options for the Haggadah templates in `haggadah-creation-panel`

image

* Changed `haggadah-view-panel` so it does not parse the Haggadah content and only shows you the Haggadah 

### test/acceptance/core.clj
* Added a function `delete-fs-emulator-data` to be used as a test fixture which takes a test and removes the data in firestore after running the test
* Changed `create-haggadah` so the Haggadah includes a timestamp field 

### test/acceptance/dashboard_test.clj
* Added `create-haggadah-test` which checks that when the user creates a Haggadah, a new Haggadah is added to firestore and the same Haggadah is shown as the first item on the user's dashboard
* Added `view-haggadot-ordered-test` which checks that when the user has a collection of Haggadot, they are displayed in order by timestamp on the dashboard

### test/acceptance/view_haggadah_test.clj
* Added `view-haggadah-with-bracha` test which checks that when the user creates a Haggadah with a bracha in it and views it from the dashboard, the bracha title and content is rendered properly
* Added `refresh-page-test` which checks that when the user views a Haggadah and refreshes the page, the user sees the Haggadah after the page has reloaded

### CODE REVIEW

File: src/haggadah/dsl.cljs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+(defn haggadah->hiccup
+  [[k {:keys [title content]}]]
+  (case k
+    :bracha [:div
+             [:div.has-text-centered.has-text-weight-bold.is-size-3.pb-2 {:data-testid :bracha-title} title]
+             [:div.has-text-right.is-size-5 {:data-testid :bracha-content }content]]
+    :else [:div]))
+
+
+#_(defn parse-haggadah
```
Comment (amirci) [inline src/haggadah/dsl.cljs]: Why leave the code?

File: src/haggadah/dsl.cljs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  Post: returns the same Haggadah represented in hiccup"
+  [content]
+  (w/walk key->hiccup #(conj [] %) #_#(into [] %) content))
+
+(defn parse-haggadah
+  "Pre: takes a Haggadah
+  Post: returns the same Haggadah represented in hiccup"
+  [content]
+  (->> content
+       ((partial map haggadah->hiccup))
```
Comment (amirci) [inline src/haggadah/dsl.cljs]: Why do you need `partial`?

File: src/haggadah/events.cljs [hunk 3]
```diff
@@ -44,6 +45,15 @@
        (.then (fn [user]  (on-success (.-user user))))
        (.catch (fn [error]  (on-error error) )))))
  
+(re-frame/reg-fx
+ ::query!
+ (fn [{:keys [path on-success on-error order-by]}]
+ (-> (firestore/instance)
```
Comment (amirci) [inline src/haggadah/events.cljs]: NIT: The event `::fetch-collection!` and `::query!` look very similar.... perhaps there's a way to extract a function?

File: src/haggadah/events.cljs [hunk 4]
```diff
@@ -64,9 +74,9 @@
 (re-frame/reg-event-fx
  ::fetch-haggadot
  (fn [{:keys [db]} [_ {:keys [on-success on-error] :or {on-error :error}}]]
-   (println "Uid: " (:uid db))
    (if (:uid db)
-     {::fetch-collection! {:path ["users" (:uid db) "haggadot"]
+     {::query! {:path ["users" (:uid db) "haggadot"]
+                :order-by (fire/orderBy "createdAt" "desc")
```
Comment (amirci) [inline src/haggadah/events.cljs]: It seems that there are many effects or events that use the convention of receiving `on-success` and/or `on-error` and then be converted using `keyword->func`. Perhaps there's a way to extract this functionality to be reused.

Comment (amirci) [reply src/haggadah/events.cljs]: Could be done in  another PR....

File: src/haggadah/views.cljs [hunk 5]
```diff
@@ -215,32 +212,42 @@ To see changes in preview edit source and then click on preview.
...
   []
-  (let [text (atom "## The best possible Haggadah")]
+  (let [haggadah (atom nil)]
     [:div.pt-4.page 
      [:div.hero.is-fullheight
       [:div.columns.is-centered
        [:div.column.is-5-tablet.is-4-desktop.is-3-widescreen
         [:form.box.mt-4 
-         [:h1.pb-4 "Please fill in the details of your Haggadah below"]
+         [:h1.pb-4 "Please make a title and select the haggadah template"]
```
Comment (amirci) [inline src/haggadah/views.cljs]: Maybe _enter a title_ is a bit more clear.

**Review by amirci — APPROVED** [review]

File: src/haggadah/views.cljs (hunk 5 above)
Comment (ebarylko) [reply src/haggadah/views.cljs]: I'll change the message so it asks for the user to enter a title.

File: src/haggadah/dsl.cljs (hunk 1 above)
Comment (ebarylko) [reply src/haggadah/dsl.cljs]: You're right. I'll take it out.

File: src/haggadah/dsl.cljs (hunk 2 above)
Comment (ebarylko) [reply src/haggadah/dsl.cljs]: After reviewing it, I see the partial is unnecessary. I'll remove it.

File: src/haggadah/events.cljs (hunk 4 above)
Comment (ebarylko) [reply src/haggadah/events.cljs]: I'll make an issue for extracting a function which takes an `on-success` and an `on-error` function and which returns the functions after applying `keyword->func` on each one.

File: src/haggadah/events.cljs (hunk 3 above)
Comment (ebarylko) [reply src/haggadah/events.cljs]: I'll see how I can extract a function from both events.

**Pull request merged**